### PR TITLE
Security Rework Part 1 of ?: Energy pistols

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

Replaces disablers with energy pistols. They're exactly the same but they have a DNA pin, and when Red Alert is enabled their kill mode is enabled. Also some nerfs to disabler beams.

## Changelog
:cl:
add: Added energy pistols. They're like a combination of disabler and laser gun, but they have a DNA pin (resettable with armory access) and kill mode is restricted to Red Alert.
tweak: Security Officers and Wardens now have energy pistols instead of disablers.
tweak: Armories now have energy pistols instead of disablers.
balance: Minor nerf to disabler beams
balance: Increase to disabler beam energy cost
tweak: Emags and EMPs now disable DNA locks.
tweak: Emagging a gun now emags its pin.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/projectile/beams.dm [hunk 1]
```diff
@@ -69,7 +69,7 @@
 /obj/item/projectile/beam/disabler
 	name = "disabler beam"
 	icon_state = "omnilaser"
-	damage = 35
+	damage = 40
```
Comment (ike709) [inline code/modules/projectiles/projectile/beams.dm]:
```suggestion
	damage = 30
```

**Comment by Sinestia** [comment]
Have any images of the sprites on hand?

Icondiff bot is well, icon diff bot.

**Comment by ike709** [comment]
@Sinestia Identical to the disabler but kill mode is red instead of blue.

**Comment by Sinestia** [comment]
> 
> 
> @Sinestia Identical to the disabler but kill mode is red instead of blue.

Ah, gotcha.

**Comment by Minelol76** [comment]
Does EMAG have any function on it and it's firing pin?

**Comment by ike709** [comment]
No.

**Comment by Minelol76** [comment]
Like, emagging it activates ability to bypass the code red restriction and removes DNA lock

**Comment by Minelol76** [comment]
Sad

**Comment by OceanFish1** [comment]
As a titanfall 2 gamer I'm getting bad vibes about "Smart-pistol"
Other than that sounds good.
Will there be a way to get disablers anyway?

File: code/modules/projectiles/guns/energy/energy_gun.dm [hunk 2]
```diff
@@ -160,3 +160,68 @@
...
+	if(dna)
+		. += "<span class='notice'>It is currently registered to: [dna]. Use an ID with armory access to reset.</span>"
+	else
+		. += "<span class='notice'>It is unregistered.</span>"
+	. += "<span class='warning'>Lethality is <b>[lethals_enabled ? "enabled" : "disabled"]</b>.</span>"
+
+/obj/item/gun/energy/smartpistol/attackby(obj/item/O, mob/user, params)
+	if(get_dna() && (ACCESS_ARMORY in O.GetAccess()))
+		to_chat(user, "<span class='notice'>You reset the DNA lock.</span>")
+		var/obj/item/firing_pin/dna/D = pin
```
Comment (PowerfulBacon) [inline code/modules/projectiles/guns/energy/energy_gun.dm:224]: Pin is not necessarily a DNA pin, as pinsc an be removed and replaced

File: code/modules/projectiles/guns/energy/energy_gun.dm [hunk 3]
```diff
@@ -160,3 +160,68 @@
...
+	if(GLOB.security_level >= SEC_LEVEL_RED && !lethals_enabled)
+		audible_message("<span class='italics'>You hear a beep from \the [name].</span>", null,  1)
+		lethals_enabled = TRUE
+		ammo_type = list(stun_ammo, kill_ammo)
+		select = 1
+		update_ammo_types()
+	else if(GLOB.security_level < SEC_LEVEL_RED && lethals_enabled)
+		audible_message("<span class='italics'>You hear a beep from \the [name].</span>", null,  1)
+		lethals_enabled = FALSE
+		ammo_type = list(stun_ammo)
```
Comment (PowerfulBacon) [inline code/modules/projectiles/guns/energy/energy_gun.dm:195]: Does lethal mode get disabled if sec level changes while on lethal mode?

File: code/modules/projectiles/projectile/beams.dm [hunk 4]
```diff
@@ -20,6 +20,9 @@
 	muzzle_type = /obj/effect/projectile/muzzle/laser
 	impact_type = /obj/effect/projectile/impact/laser
 
+/obj/item/projectile/beam/laser/smartpistol
+	damage = 30
```
Comment (PowerfulBacon) [inline code/modules/projectiles/projectile/beams.dm]: Buffed lasers for the smart pistol only?

**Comment by Mat05usz** [comment]
In my opinion it's an unnecessary PR, buffing security and nerfing everyone else. Why the hell redshits are given eguns roundstart (red alert isn't a problem to raise for even very minor reasons) while also restricting the use of this gun from anyone else?

**Comment by ItsTisme** [comment]
>(red alert isn't a problem to raise for even very minor reasons)

On sage atleast there are rules on when you can and cant raise alert levels. So you cant just make it red alert round start unless you enjoy eating bwoink.

That being said. Maby just giving them mindshield pins would be more interesting?

**Comment by Naevii** [comment]
What exactly is that "minor nerf" to beams and how much cost is increased?

**Comment by AnCopper** [comment]
> What exactly is that "minor nerf" to beams and how much cost is increased?

Energy per shot increased to 60
Damage per shot reduced to 30 (from 35)

File: code/modules/projectiles/guns/energy/energy_gun.dm [hunk 5]
```diff
@@ -160,3 +160,68 @@
...
+		select = 1
+		update_ammo_types()
+		select_fire()
+
+/obj/item/gun/energy/smartpistol/proc/get_dna()
+	var/obj/item/firing_pin/dna/D = pin
+	if(D.unique_enzymes)
+		return D.unique_enzymes
+	return null
```
Comment (Archanial) [inline code/modules/projectiles/guns/energy/energy_gun.dm]:
```suggestion
	return D.unique_enzymes ? D.unique_enzymes : null
```

**Review by Archanial — COMMENTED** [review]
not sure if locking it to dna is good idea

**Comment by Archanial** [comment]
Also attacking it with emag should reset DNA lock.

**Comment by Mat05usz** [comment]
> > (red alert isn't a problem to raise for even very minor reasons)
> 
> On sage atleast there are rules on when you can and cant raise alert levels. So you cant just make it red alert round start unless you enjoy eating bwoink.

Yeah there are rules, which this PR gives an incentive to break more often thus putting even more strain on our already undermanned admin team.

**Comment by Sinmero** [comment]
'minor nerf' (33%) moment

**Comment by Kapu1178** [comment]
Honestly this is the complete wrong direction to buff security. Roundstart lethals is bad. Adding arguably artificial barriers to using a common item (the disabler) is bad. I think a significantly simpler and better way for all parties to buff security would be improvements to the disabler's non-lethal takedowns. Currently, disablers are very fucking awful to use. The projectile is slow, takes 10 years to knock someone down, and you have another option thats better in every way: the energy bola.


An increase in the disabler's projectile speed and a buff to the stamina damage dealt is more elegant, more effective, and generally just better for the game.

**Comment by Archanial** [comment]
> An increase in the disabler's projectile speed and a buff to the stamina damage dealt is more elegant, more effective, and generally just better for the game.

?

**Comment by Isy232** [comment]
> ## About The Pull Request
> Replaces disablers with smart pistols. They're exactly the same but they have a DNA pin, and when Red Alert is enabled their kill mode is enabled. Also some nerfs to disabler beams.
> 
> ## Changelog
> 🆑
> add: Added smart pistols. They're like a combination of disabler and laser gun, but they have a DNA pin (resettable with armory access) and kill mode is restricted to Red Alert.
> tweak: Security Officers and Wardens now have smart pistols instead of disablers.
> tweak: Armories now have smart pistols instead of disablers.
> balance: Minor nerf to disabler beams
> balance: Increase to disabler beam energy cost
> /🆑

Lethal mode should be set to blue alert.

As per SOP in some instances lethal force is permitted on blue alert.

**Comment by Mat05usz** [comment]
> Honestly this is the complete wrong direction to buff security. Roundstart lethals is bad. Adding arguably artificial barriers to using a common item (the disabler) is bad. I think a significantly simpler and better way for all parties to buff security would be improvements to the disabler's non-lethal takedowns. Currently, disablers are very fucking awful to use. The projectile is slow, takes 10 years to knock someone down, and you have another option thats better in every way: the energy bola.
> 
> An increase in the disabler's projectile speed and a buff to the stamina damage dealt is more elegant, more effective, and generally just better for the game.

Any direction to buff security is wrong, because they don't need buffs. Also disabler is one of the best non lethal weapons lol

>Lethal mode should be set to blue alert.

>As per SOP in some instances lethal force is permitted on blue alert.

No, even with those guns locked behind red alert it's still bad, as essentially it means that every sec officer is walking around with personal egun, that cannot be used against him.

**Comment by DatBoiTim** [comment]
> As a titanfall 2 gamer I'm getting bad vibes about "Smart-pistol"
> Other than that sounds good.
> Will there be a way to get disablers anyway?

Glad I’m not the only one who the knee jerk “This literally adds aimbot doesn’t it?”

File: code/modules/projectiles/guns/energy/energy_gun.dm (hunk 2 above)
Comment (ike709) [reply code/modules/projectiles/guns/energy/energy_gun.dm:224]: The var to remove it is false by default.

File: code/modules/projectiles/guns/energy/energy_gun.dm (hunk 3 above)
Comment (ike709) [reply code/modules/projectiles/guns/energy/energy_gun.dm:195]: Sure does.

File: code/modules/projectiles/projectile/beams.dm (hunk 4 above)
Comment (ike709) [reply code/modules/projectiles/projectile/beams.dm]: Yep.

**Comment by ike709** [comment]
Emags can now override the DNA lock.

File: code/modules/projectiles/pins.dm [hunk 6]
```diff
@@ -171,7 +172,7 @@
 			to_chat(user, "<span class='notice'>DNA-LOCK SET.</span>")
 
 /obj/item/firing_pin/dna/pin_auth(mob/living/carbon/user)
-	if(unique_enzymes && user?.dna?.unique_enzymes == unique_enzymes) //First check is in case they're both null
+	if(unique_enzymes && ((obj_flags & EMAGGED) || user?.dna?.unique_enzymes == unique_enzymes)) //First check is in case they're both null
```
Comment (Archanial) [inline code/modules/projectiles/pins.dm:175]: If it was emagged but didn't have unique_enzymes that would still return FALSE? Why?

**Review by Archanial — COMMENTED** [review]
Does emagging pistol directly work? If not it should. Also, if you really want to keep the ID lock on please don't force every single antagonist to buy emag. Add some way to manually hack the mechanism (perhaps with tools).

**Comment by ike709** [comment]
> Add some way to manually hack the mechanism (perhaps with tools).

No.

File: code/modules/projectiles/pins.dm (hunk 6 above)
Comment (ike709) [reply code/modules/projectiles/pins.dm:175]: To simplify things. Pin auth failure with null `unique_enzymes` sets the var.

**Comment by DatBoiTim** [comment]
> Add some way to manually hack the mechanism (perhaps with tools).

Can't you remove firing pins?

**Comment by snek-with-tophat** [comment]
Maybe make emps reset the stored dna? 
That way there is at least some way for it to be unlocked without traitor items

**Comment by Archanial** [comment]
> 
> 
> > Add some way to manually hack the mechanism (perhaps with tools).
> 
> No.

let's force antags to buy emag ok
just remove the whole ID lock, mindshield pins are locked behind research for a reason.

**Comment by ike709** [comment]
EMPs now reset the lock, since that's a good idea.

File: code/modules/projectiles/guns/energy/energy_gun.dm [hunk 7]
```diff
@@ -227,3 +227,10 @@
 			D.obj_flags &= ~EMAGGED
 		investigate_log("dna lock reset by [key_name(user)]", INVESTIGATE_RECORDS)
 	..()
+
+/obj/item/gun/energy/smartpistol/emp_act(severity)
+	audible_message("<span class='italics'>You hear erratic beeping from \the [name].</span>", null,  1)
+	var/obj/item/firing_pin/dna/D = pin
+	D.unique_enzymes = null
```
Comment (Archanial) [inline code/modules/projectiles/guns/energy/energy_gun.dm:234]: this will runtime if gun doesn't have pin

Comment (ike709) [reply code/modules/projectiles/guns/energy/energy_gun.dm:234]: If pin code is correct the pin can't be removed.

Comment (Archanial) [reply code/modules/projectiles/guns/energy/energy_gun.dm:234]: imo better be safe

**Comment by utilbird** [comment]
DNA pin is a bit silly, especially when the only (practical) counter is to buy the already extremely versatile and borderline necessary emag. 
You haven't even provided a reason as to why this is a good change in the pr description.


smart pistol is also a bad name. I'd just keep it the same or change it to "energy pistol" if you really needed to

**Comment by snek-with-tophat** [comment]
DNA pins sound like ERT level equipment rather then standard issue round-start gear, especially when mindshield pins, which can perform the same function, are locked behind a good amount of research and a hefty material price

**Comment by WhyIsCaeciliusTaken** [comment]
> DNA pin is a bit silly, especially when the only (practical) counter is to buy the already extremely versatile and borderline necessary emag.
> You haven't even provided a reason as to why this is a good change in the pr description.
> 
> smart pistol is also a bad name. I'd just keep it the same or change it to "energy pistol" if you really needed to

It's good for the game because now random assistants can't validhunt as effectively as sec.

**Comment by Archanial** [comment]
> 
> 
> > DNA pin is a bit silly, especially when the only (practical) counter is to buy the already extremely versatile and borderline necessary emag.
> > You haven't even provided a reason as to why this is a good change in the pr description.
> > smart pistol is also a bad name. I'd just keep it the same or change it to "energy pistol" if you really needed to
> 
> It's good for the game because now random assistants can't validhunt as effectively as sec.

Just get one that isn't DNA bound yet and bind it to yourself.

**Comment by DatBoiTim** [comment]
> > > You haven't even provided a reason as to why this is a good change in the pr description.

Good point, let's hear it.

> > > smart pistol is also a bad name. I'd just keep it the same or change it to "energy pistol" if you really needed to

Yes, smart pistol has some connotations...

> > It's good for the game because now random assistants can't validhunt as effectively as sec.
> 
> 
> 
> Just get one that isn't DNA bound yet and bind it to yourself.

Or use an EMP on one.

**Comment by Sinestia** [comment]
> 
> 
> > > (red alert isn't a problem to raise for even very minor reasons)
> > 
> > 
> > On sage atleast there are rules on when you can and cant raise alert levels. So you cant just make it red alert round start unless you enjoy eating bwoink.
> 
> Yeah there are rules, which this PR gives an incentive to break more often thus putting even more strain on our already undermanned admin team.

Rules like these are easy to enforce, that's fine. 

A lot of the strain is from lofty RP rules that require either A. someone to ahelp it or B. a jannies to be extremely active in finding them, creating a cycle of people thinking they don't actually matter because they never see them enforced.

 This however, is very blatant, and red alerting usually has extremely clear chat logs that mean you won't even have to bwoink for.

**Comment by ike709** [comment]
> people think this is a buff

wew lad

Also the point of this is because part 2 of the ongoing sec rework is going to be significantly reducing security's available arsenal during normal station operations. So I wanted security to have a restricted lethals option despite that.

I will rename it from smart pistol though since too many people play Titanfall.

**Comment by yyzsong** [comment]
> EMPs now reset the lock, since that's a good idea.

Not in changelog smh smh

**Comment by PowerfulBacon** [comment]
> 
> 
> > people think this is a buff
> 
> wew lad

People think its a buff because its a backpack-sized energy gun with the most powerful laser ammo type in the game, the lasers deal 50% more damage than regular lasers and consume less ammo.
You can fit multiple laser guns in your back by setting the alert to red and picking a bunch of these up.

edit: nvm it doesn't fit in backpacks, but its still a buffed laser gun and nerfed disabler combined into 1 that requires red alert to use the lasers

**Comment by utilbird** [comment]
Why doesn't a pistol fit in a backpack

**Comment by Raven-Industries** [comment]
I do like the idea, although the disabler nerf is a bit annoying; it's already hard enough to hit a moving target enough times to drop them (especially when the beam is so slow), now we're getting fewer shots and lower power per shot.

**Comment by Archanial** [comment]
> 
> 
> Why doesn't a pistol fit in a backpack

no powercreep

**Comment by utilbird** [comment]
> no powercreep

It's not powercreep if pistols already fit in backpacks

**Comment by eeSPee** [comment]
I don't know. It already is a unnecessary buff to a powerful ranged weapon. The disabler, a round start weapon available to all officers, kinda makes all energy guns obsolete anyway. Now if it can turn into an egun at will and starts with a DNA lock, what even is the point of energy guns?

Never really had a problem that my disabler was stolen by random assistants as a sec officer to begin with, so there's no meta it puts a stop to. It's cool and it makes sense in a futuristic setting, but it just late game tech without any downsides.

**Comment by BrumalLake** [comment]
This is good, it gives stealth antags without stamina armor a bit more of a fighting chance on green and blue alert, and gives sec a bit more of a fighting chance against heavily armored and mass conversion antags like cult and revs which are pretty much when red alert should be raised
Also punishes antags for going too loud

**Comment by eeSPee** [comment]
I don't think this is a good change for the metagame.

**Comment by utilbird** [comment]
whether it's good for the game or not is the only bit that really matters, the metagame will always change

Conforming to the metagame is very restrictive design-wise too

**Comment by PowerfulBacon** [comment]
I really dont like the DNA lock personally

**Comment by ike709** [comment]
The only people who should be looting sec officers for weapons have the means to bypass the DNA lock.

**Comment by eLeCtrOssSnake** [comment]
Now half of the whole sec gear becomes useless because you now apparently need to spend tc's on emag to make it useful. Other non-uplink antags have no reason to loot sec. Never saw an issue with security, but I think this just introduces it. Also, you can now get robusted by the meth guy because he's stamina damage immune and you have no lethals because hurr durr smart red-alert gun. It seems very off. These changes lead to a huge change in the game, which I don't think is required nor justified.

What was the issue anyway?

Kind of an MRP-only rework again.

>I don't think this is a good change for the metagame.
Exactly. It adds too many strange constraints that affect security way too much.

**Comment by eLeCtrOssSnake** [comment]
I would personally nerf double-wielded disablers because this is kind of OP, even on metagame level.

File: code/modules/projectiles/projectile/beams.dm [hunk 8]
```diff
@@ -20,6 +20,9 @@
 	muzzle_type = /obj/effect/projectile/muzzle/laser
 	impact_type = /obj/effect/projectile/impact/laser
 
+/obj/item/projectile/beam/laser/energypistol
+	damage = 30
```
Comment (eeSPee) [inline code/modules/projectiles/projectile/beams.dm]: For refference, egun does 15 damage, and laser does 20.

Who do the non-lethal hand held pistols deal twice that amount?

Comment (yorii) [reply code/modules/projectiles/projectile/beams.dm]: holy fuck, @ike709 are you crazy?

**Comment by Lagomorphica** [comment]
Hello. I would just like to comment on some major oversights with this that could be detrimental in a full merge. Currently, with testing on my own private server with this PR on master branch, the energy pistols have roughly 16 lethal shots on Red Alert. This allows for a total of 32 shots if you dual wield them, and they cannot be stolen by antagonists short of imitation of a Security Officer (requires stealing a mindshield), EMP, or an emag being used.

While this is not itself the biggest issue, eguns only get 10 shots, and laser guns have 14, placing the energy pistol ahead of the curve of all other laser weapons, including the Head of Securities gun.

In addition to this, the laser projectile itself does 30 damage - lasers fired from all other station-available gear only do 20. The energy pistols, as of now, do as much burn damage as an emitter firing, and can be dual wielded.

This also enables them to husk targets with one energy pistol, and gives a maximum damage value, on an unarmoured target, of 480 burn.

**Comment by ike709** [comment]
> Someone finally did the math for balancing

Wonderful, now I don't have to do it myself.

**Comment by AnCopper** [comment]
Mounting a seclite is broken on the energy pistol.

**Comment by BrumalLake** [comment]
I think around 25-26 damage for all lethals including the laser pistol's would make people stop comparing this gun to others in the armory
Especially since some people falsely believe that lasers do 25 damage right now so they wouldn't be able to argue

**Comment by eeSPee** [comment]
> I think around 25-26 damage for all lethals including the laser pistol's would make people stop comparing this gun to others in the armory
> Especially since some people falsely believe that lasers do 25 damage right now so they wouldn't be able to argue

And COMPLETELY throw balance out the window?

Right now, the energy gun deals 15 damage, laser gun does 20 damage and laser pistol deals a whopping 30 damage.

The unnecessary buff to energy pistols already makes antags like revs and xenos a joke; no more threatening than 10 monkey cubes. Heck, a xenomorph queen is no longer a threat against a single wielded energy pistol on red alert.

File: code/modules/projectiles/projectile/beams.dm (hunk 4 above)
Comment (TeomanTheGreat) [reply code/modules/projectiles/projectile/beams.dm]: Ike has yet to learn that 30 burn damage per hit is a lot, apparently.

**Comment by TeomanTheGreat** [comment]
Here is a forum thread discussing the issues with this pr that someone made. I have no real authority when I say this but i believe this should be looked at before this pr is merged.  https://forums.beestation13.com/t/energy-pistols-fuck-up-everything/

**Comment by DatBoiTim** [comment]
> Here is a forum thread discussing the issues with this pr that someone made. I have no real authority when I say this but i believe this should be looked at before this pr is merged. https://forums.beestation13.com/t/energy-pistols-fuck-up-everything/

If that comment about soloing blobs is true, the pistols either need nerfs or literally everything needs a rebalance.

File: code/modules/projectiles/ammunition/energy/laser.dm [hunk 9]
```diff
@@ -10,6 +10,11 @@
 	e_cost = 71
 	select_name = "kill"
 
+/obj/item/ammo_casing/energy/lasergun/energypistol
+	projectile_type = /obj/item/projectile/beam/laser/energypistol
+	e_cost = 60
```
Comment (ike709) [inline code/modules/projectiles/ammunition/energy/laser.dm]:
```suggestion
	e_cost = 80
```

File: code/modules/projectiles/projectile/beams.dm (hunk 8 above)
Comment (ike709) [inline code/modules/projectiles/projectile/beams.dm]:
```suggestion
	damage = 20
```

**Comment by ike709** [comment]
The rather insanely powerful lethal mode has been nerfed.

**Comment by TeomanTheGreat** [comment]
I no longer have any problems with this pr. Thanks.

**Comment by DatBoiTim** [comment]
Now we just got figure out what the fuck to do with the red alert PR

**Comment by utilbird** [comment]
How does the DNA lock and laser mode make the game better. Laser mode harms gadget diversity at best and makes red alert a metagame nightmare at worst. DNA lock is another can of worms so I'm just going to rip my short response from the forum

> Smartguns are silly.
Seccies should be punished for losing their gun no matter how it happened, if you really wanted to change something about that you’d focus on making it harder to steal guns rather than tampering with the reason to do so

> Good balance decisions address opportunities, not outcomes

**Comment by ike709** [comment]
> How does the DNA lock and laser mode make the game better.

The former is because anybody who *should* be stealing sec guns in the first place have the means to bypass it. The latter is because I intend to heavily restrict security's access to lethals in general across a series of PRs, which if my summer research ever ends I'll start making progress on again.

**Comment by AutonomousAnonymous** [comment]
> anybody who _should_ be stealing sec guns in the first place have the means to bypass it 

only antags with access to emags and changelings who can change their dna to match the lock on the fly can bypass it, blood bros, heretics, cultists, revs and almost every single midround barring the loneop (who, with access to better lethal and nonlethal weapons anyway, really has no reason to be using the dinky little tic tacs in the first place) can pretty much get fucked

**Comment by ike709** [comment]
EMPs

**Comment by AutonomousAnonymous** [comment]
i feel like that should be on the changelog then

**Comment by PowerfulBacon** [comment]
Personally I dont really like the DNA locks anyway. Also has the laser damage been rebalanced yet?

File: code/modules/projectiles/projectile/beams.dm [hunk 10]
```diff
@@ -20,6 +20,9 @@
 	muzzle_type = /obj/effect/projectile/muzzle/laser
 	impact_type = /obj/effect/projectile/impact/laser
 
+/obj/item/projectile/beam/laser/energypistol
+	damage = 20
```
Comment (PowerfulBacon) [inline code/modules/projectiles/projectile/beams.dm:24]: there is no point for this, its litreally just a normal laser

**Comment by PowerfulBacon** [comment]
Is this ready?

**Comment by ike709** [comment]
Supposedly there's an issue with seclites that I need to look into when I have time.

**Comment by ike709** [comment]
I don't have time for this and I lost all motivation to nerf sec.

**Pull request closed without merging**